### PR TITLE
Combine mstore_general and mload_general into one flag

### PR DESCRIPTION
**Pull request by LindaGuiga**
This PR merges the `mstore_general` and `mload_general` flags into one. Since the flags opcodes do not abide by the `decode.rs` requirements, the decoding checks are done manually in the constraints.

### CODE REVIEW

File: evm/src/cpu/decode.rs [hunk 1]
```diff
@@ -55,8 +56,6 @@ const OPCODES: [(u8, usize, bool, usize); 32] = [
     (0xf6, 0, true, COL_MAP.op.get_context),
     (0xf7, 0, true, COL_MAP.op.set_context),
     (0xf9, 0, true, COL_MAP.op.exit_kernel),
-    (0xfb, 0, true, COL_MAP.op.mload_general),
```
Comment (Nashtare) [inline evm/src/cpu/decode.rs:58]: I'd add a comment for clarity (maybe like what's been done above for logic flags)

File: evm/src/cpu/decode.rs [hunk 2]
```diff
@@ -141,6 +144,9 @@ pub fn eval_packed_generic<P: PackedField>(
     let flag = lv.op.logic_op;
     yield_constr.constraint(cycle_filter * flag * (flag - P::ONES));
 
+    let m_op_flag = lv.op.m_op_general;
+    yield_constr.constraint(cycle_filter * m_op_flag * (m_op_flag - P::ONES));
+
     // Now check that they sum to 0 or 1.
     // Includes the logic_op flag encompassing AND, OR and XOR opcodes.
```
Comment (Nashtare) [inline evm/src/cpu/decode.rs:151]: Update comment to take into account the new flag perhaps?

File: evm/src/cpu/decode.rs [hunk 3]
```diff
@@ -221,12 +242,18 @@ pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
     let constr = builder.mul_extension(cycle_filter, constr);
     yield_constr.constraint(builder, constr);
 
+    let m_op_flag = lv.op.m_op_general;
+    let constr = builder.mul_sub_extension(m_op_flag, m_op_flag, m_op_flag);
+    let constr = builder.mul_extension(cycle_filter, constr);
+    yield_constr.constraint(builder, constr);
+
     // Now check that they sum to 0 or 1.
     // Includes the logic_op flag encompassing AND, OR and XOR opcodes.
```
Comment (Nashtare) [inline evm/src/cpu/decode.rs:251]: Same comment as above.

File: evm/src/cpu/memio.rs [hunk 4]
```diff
@@ -38,14 +41,47 @@ fn eval_packed_load<P: PackedField>(
     for &channel in &lv.mem_channels[4..NUM_GP_CHANNELS - 1] {
         yield_constr.constraint(filter * channel.used);
     }
+
+    // Stack behavior constraints
```
Comment (Nashtare) [inline evm/src/cpu/memio.rs:45]: Can't we mimic what's been done in #1184 to remove all this code duplication and make it clearer?

File: evm/src/witness/transition.rs [hunk 5]
```diff
@@ -185,8 +185,8 @@ fn fill_op_flag<F: Field>(op: Operation, row: &mut CpuColumnsView<F>) {
         Operation::GetContext => &mut flags.get_context,
         Operation::SetContext => &mut flags.set_context,
         Operation::ExitKernel => &mut flags.exit_kernel,
-        Operation::MloadGeneral => &mut flags.mload_general,
-        Operation::MstoreGeneral => &mut flags.mstore_general,
+        Operation::MloadGeneral => &mut flags.m_op_general,
+        Operation::MstoreGeneral => &mut flags.m_op_general,
```
Comment (Nashtare) [inline evm/src/witness/transition.rs:189]: Same comment as in #1184, could be combined to make flags appear only once on the right side
```suggestion
        Operation::MloadGeneral | Operation::MstoreGeneral => &mut flags.m_op_general,
```

**Review by Nashtare — COMMENTED** [review]
Looks good to me. Would be nice if you could tag someone from Zero as well for a cross-team review!

**Comment by LindaGuiga** [comment]
@wborgeaud Would you have time to review this by any chance, please? 🙏🏻

File: evm/src/cpu/decode.rs [hunk 6]
```diff
@@ -99,6 +100,10 @@ pub fn generate<F: RichField>(lv: &mut CpuColumnsView<F>) {
         let flag = available && opcode_match;
         lv[col] = F::from_bool(flag);
     }
+
+    if opcode == 0xfb || opcode == 0xfc {
+        lv[COL_MAP.op.m_op_general] = F::from_bool(kernel);
```
Comment (nbgl) [inline evm/src/cpu/decode.rs:105]:
```suggestion
        lv.op.m_op_general = F::from_bool(kernel);
```

File: evm/src/cpu/decode.rs [hunk 7]
```diff
@@ -173,6 +178,20 @@ pub fn eval_packed_generic<P: PackedField>(
...
     }
+
+    // Manually check lv.op.m_op_constr
+    let opcode: P = lv
+        .opcode_bits
+        .into_iter()
+        .enumerate()
+        .map(|(i, bit)| bit * P::Scalar::from_canonical_u64(1 << i))
+        .sum();
+    yield_constr.constraint((P::ONES - kernel_mode) * lv[COL_MAP.op.m_op_general]);
```
Comment (nbgl) [inline evm/src/cpu/decode.rs:189]:
```suggestion
    yield_constr.constraint((P::ONES - kernel_mode) * lv.op.m_op_general);
```

File: evm/src/cpu/decode.rs [hunk 8]
```diff
@@ -173,6 +178,20 @@ pub fn eval_packed_generic<P: PackedField>(
...
+
+    let m_op_constr = (opcode - P::Scalar::from_canonical_usize(0xfb_usize))
+        * (opcode - P::Scalar::from_canonical_usize(0xfc_usize))
+        * lv[COL_MAP.op.m_op_general];
```
Comment (nbgl) [inline evm/src/cpu/decode.rs:193]:
```suggestion
        * lv.op.m_op_general;
```

File: evm/src/cpu/decode.rs [hunk 9]
```diff
@@ -251,4 +270,28 @@ pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
...
+        .fold(builder.zero_extension(), |cumul, bit| {
+            builder.mul_const_add_extension(F::TWO, cumul, bit)
+        });
+
+    let mload_opcode = builder.constant_extension(F::Extension::from_canonical_usize(0xfb_usize));
+    let mstore_opcode = builder.constant_extension(F::Extension::from_canonical_usize(0xfc_usize));
+
+    let one_extension = builder.constant_extension(F::Extension::ONE);
+    let is_not_kernel_mode = builder.sub_extension(one_extension, kernel_mode);
+    let constr = builder.mul_extension(is_not_kernel_mode, lv[COL_MAP.op.m_op_general]);
```
Comment (nbgl) [inline evm/src/cpu/decode.rs:288]:
```suggestion
    let constr = builder.mul_extension(is_not_kernel_mode, lv.op.m_op_general);
```

File: evm/src/cpu/decode.rs [hunk 10]
```diff
@@ -251,4 +270,28 @@ pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
...
+    yield_constr.constraint(builder, constr);
+
+    let mload_constr = builder.sub_extension(opcode, mload_opcode);
+    let mstore_constr = builder.sub_extension(opcode, mstore_opcode);
+    let mut m_op_constr = builder.mul_extension(mload_constr, mstore_constr);
+    m_op_constr = builder.mul_extension(m_op_constr, lv[COL_MAP.op.m_op_general]);
```
Comment (nbgl) [inline evm/src/cpu/decode.rs:294]:
```suggestion
    m_op_constr = builder.mul_extension(m_op_constr, lv.op.m_op_general);
```

**Review by nbgl — APPROVED** [review]
Looks good!! Just a few nits

**Pull request merged**